### PR TITLE
Create JSON schema test cases

### PR DESCRIPTION
**Pull request by JiwoonKim**
**Closes Issue:** #3 
create general schemas and edge cases for testing purposes

**Changes Applied**
created following folders and files within `/schemas` folder

- [x] `/basic data types`

   - [x] string
   - [x] numeric
   - [x] boolean
   - [x] null
   - [x] array
   - [x] object

- [x] `/ref types`
   - [x] reference $def
   - [x] circular reference

- [x] `/combination types`
    - [x] allOf
    - [x] oneOf
    - [x] anyOf
    - [x] not

- [x] `/demo` (schemas used in Taskcluster)
    - [x] basic schema (basic data types, arrays, object)
    - [x] combination schema
    - [x] references #definition schema
    - [x] circular referencing schema

### CODE REVIEW

File: schemas/basicDataTypes/array/additionalItems.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    "type": "array",
+    "items": [
+      {
+        "type": "number"
+      },
+      {
+        "type": "string"
+      }
+    ],
+    "additionalItems": { "type": "boolean" }
```
Comment (JiwoonKim) [inline schemas/basicDataTypes/array/additionalItems.json]: Since `additionalItems` keyword is meaningless with list validation type of lists
https://json-schema.org/understanding-json-schema/reference/array.html#items,
used the keyword with tuple validation example instead!

File: schemas/basicDataTypes/array/containsValidation.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  "type": "array",
+  "contains": {
+    "type": "number"
+  }
```
Comment (JiwoonKim) [inline schemas/basicDataTypes/array/containsValidation.json]: Specified in draft 6, `contains` keyword is defined as

> while the items schema must be valid for every item in the array, the contains schema only needs to validate against one or more items in the array. 

I included a test case of this since it is specified in the JSON Schema docs,
but I am not sure how this should be displayed in our schema viewer at the moment.
Should this also be a case we should consider testing or drop completely?

File: schemas/basicDataTypes/object/propertySpecifications.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  },
+  "patternProperties": {
+    "^a": { "type": "string" },
+    "^b": { "type": "integer" }
+  },
+  "additionalProperties": {
+    "type": "boolean"
+  },
+  "minProperties": 1,
+  "maxProperties": 10
```
Comment (JiwoonKim) [inline schemas/basicDataTypes/object/propertySpecifications.json]: I combined all the specifications of 
`propertyNames`, `patternProperties`, `additionalProperties` and size specs of `minProperties` & `maxProperties` 

all in ONE file since these would be listed together in the same section when displayed in the viewer.

P.S just in case anyone is curious, a valid JSON that passes this schema would be
```json
{
  "apple": "red fruit",
  "black": 255,
  "catty": true 
}
```

**Comment by JiwoonKim** [comment]
@helfi92 
Added all the test cases listed above!

Most of them are based on the [json-schema documentation](https://json-schema.org/understanding-json-schema/) 
so that it's easy to reference from the docs when necessary :)

I've tried to include all of the keywords that can be used
and have arranged them in the `/schema` folder according to their USES.
For example,

- some property keywords may be joined together like in [`/object/propertySpecifications.json`](https://github.com/taskcluster/material-ui-json-schema-viewer/pull/5/files/e355e7d1fa136b75a26408e7a81976c7251c58b0#diff-655250479660dd299e04e9a496128093)
OR
- some may be separated to test different cases like in [`array/additionalItems.json`](https://github.com/taskcluster/material-ui-json-schema-viewer/pull/5/files/e355e7d1fa136b75a26408e7a81976c7251c58b0#diff-5f0705e8fd393827213cbff44286c299) and [`/array/tupleValidation.json `](https://github.com/taskcluster/material-ui-json-schema-viewer/pull/5/files/e355e7d1fa136b75a26408e7a81976c7251c58b0#diff-953a74ce8342d31086e32e65631a6176)

Let me know if there is anything to change!

File: schemas/basicDataTypes/array/containsValidation.json (hunk 2 above)
Comment (JiwoonKim) [reply schemas/basicDataTypes/array/containsValidation.json]: Found that Taskcluster actually uses the `contains` keywords in their schemas, so should definitely test this as well!

`metaschema.json` uses
```json
"contains": {
     "$ref": "#/definitions/schema"
},
```

Comment (helfi92) [reply schemas/basicDataTypes/array/containsValidation.json]: Good find. Yes, it's a good idea to have this tested.

**Comment by djmitche** [comment]
This is awesome!  Are there plans to involve these in some automated (`yarn test`) test cases, too?

**Comment by JiwoonKim** [comment]
> This is awesome! Are there plans to involve these in some automated (`yarn test`) test cases, too?

It would be great if the test cases could be automated!
The original plan is to use the test cases in order to check how different schema cases would layout in the viewer UI. And also to make sure all the keywords are displayed as well.

I'm not familiar with automating tests to check UIs, but if it is possible, this could be something neat to use when developing the viewer. 

@helfi92 Is it possible to automate tests for UIs or should we just stick to checking the schemas layouts manually?

**Comment by helfi92** [comment]
> @helfi92 Is it possible to automate tests for UIs or should we just stick to checking the schemas layouts manually?

We could use the snapshot feature in jest to make sure the UI doesn't change unexpectedly. https://jestjs.io/docs/en/snapshot-testing. We use this feature in https://github.com/mozilla-frontend-infra/components for example.
